control-plane/agents/core/src/pool/specs.rs

```diff
@@ -0,0 +1,450 @@
+                    if pool_spec.updating {
+                        // it's already being created
+                        return Err(SvcError::Conflict {});
+                    } else if pool_spec.state.creating() {
+                        // this might be a retry, check if the params are the
+                        // same if so, let's retry!
+                        if pool_spec.ne(request) {
+                            // if not then we can't proceed, so signal a
+                            // conflict
+                            return Err(SvcError::Conflict {});
+                        }
+                    } else {
+                        return Err(SvcError::AlreadyExists {
+                            kind: ResourceKind::Pool,
+                            id: request.id.to_string(),
+                        });
+                    }
+                    pool_spec.updating = true;
+                }
+                spec
+            } else {
+                let spec = PoolSpec::from(request);
+                // write the spec to the persistent store
+                {
+                    let mut store = registry.store.lock().await;
+                    store.put_obj(&spec).await?;
+                }
+                // add spec to the internal spec registry
+                let spec = Arc::new(Mutex::new(spec));
+                specs.pools.insert(request.id.clone(), spec.clone());
+                spec
+            }
+        };
+
+        let result = node.create_pool(request).await;
+        let mut pool_spec = pool_spec.lock().await;
+        pool_spec.updating = false;
+        if result.is_ok() {
+            let mut pool = pool_spec.clone();
+            pool.state = PoolSpecState::Created(PoolState::Online);
+            let mut store = registry.store.lock().await;
+            store.put_obj(&pool).await?;
+            pool_spec.state = PoolSpecState::Created(PoolState::Online);
+        } else {
+            drop(pool_spec);
+            self.del_pool(&request.id).await;
+            let mut store = registry.store.lock().await;
+            let _ = store.delete_kv(&request.id.key()).await;
+        }
+
+        result
+    }
+
+    pub(crate) async fn destroy_pool(
+        &self,
+        registry: &Registry,
+        request: &DestroyPool,
+    ) -> Result<(), SvcError> {
+        // what if the node is never coming back?
+        // do we need a way to forcefully "delete" things?
+        let node = registry.get_node_wrapper(&request.node).await.context(
+            NodeNotFound {
+                node_id: request.node.clone(),
+            },
+        )?;
+
+        let pool_spec = self.get_pool(&request.id).await;
+        if let Some(pool_spec) = &pool_spec {
+            let mut pool_spec = pool_spec.lock().await;
+            if pool_spec.updating {
+                return Err(SvcError::Conflict {});
+            } else if pool_spec.state.deleted() {
+                return Ok(());
```

Comment:
This is the case where between getting the resource from the list and trying to delete it it "got deleted" by another call that managed to get in between. I guess we massage it into a not found error since the end result the same as never finding it the first place?